File: graph_algorithms/LongestFlightRoute.java
import java.util.*;
import java.io.*;



public class LongestFlightRoute {

    public static void main(String[] args) {
        FastScanner sc = new FastScanner();
        int n = sc.nextInt();
        int m = sc.nextInt();

        List<List<Integer>> graph = new ArrayList<>();
        for(int i=0; i<=n; i++) {
            graph.add(new ArrayList<>());
        }

        for(int i=0; i<m; i++) {
            int a = sc.nextInt();
            int b = sc.nextInt();
            graph.get(a).add(b);
        }

        int dist[] = new int[n+1];
        int forwardNode[] = new int[n+1];
        for(int i=0; i<=n; i++) {
            dist[i] = -1;
            forwardNode[i] = -1;
        }
        dfs(graph, dist, forwardNode, 1, n);
        
        if(dist[1] == -1) {
            System.out.println("IMPOSSIBLE");
            return;
        }

        StringBuilder s = new StringBuilder();
        s.append(dist[1] + "\n");
        int index = 1;
        while(index != n) {
            s.append(index + " ");
            index = forwardNode[index];
        }
        s.append(n);
        System.out.println(s);
    }

    public static int dfs(List<List<Integer>> graph, int[] dist, int[] forwardNode, int node, int dest) {
        if(node == dest) {
            dist[node] = 1;
            forwardNode[node] = 0;
            return 1;
        }

        if(dist[node] != -1) {
            return dist[node];
        }

        int distFromDest = -1;
        int nextNode = -1;
        for(int neighbor : graph.get(node)) {
            int d = dfs(graph, dist, forwardNode, neighbor, dest);
            if(d == -1) {
                continue;
            }
            if(d+1 > distFromDest) {
                distFromDest = d+1;
                nextNode = neighbor;
            }
        }

        dist[node] = distFromDest;
        forwardNode[node] = nextNode;
        return distFromDest;
    }
    
}

class FastScanner {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer("");
 
    String next() {
        while (!st.hasMoreTokens())
            try {
                st = new StringTokenizer(br.readLine());
            } catch (IOException e) {
                e.printStackTrace();
            }
        return st.nextToken();
    }
 
    int nextInt() {
        return Integer.parseInt(next());
    }
 
    int[] readArray(int n) {
        int[] a = new int[n];
        for (int i = 0; i < n; i++)
            a[i] = nextInt();
        return a;
    }
 
    long nextLong() {
        return Long.parseLong(next());
    }
}

File: graph_algorithms/longestflightroute.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>

using namespace std;

void dfs(int node, vector<vector<int>>& graph, vector<int>& dist, vector<int>& forwardNode, vector<bool>& visited, int dest) {
    visited[node] = true;
    if (node == dest) {
        dist[node] = 1;
        forwardNode[node] = 0;
        return;
    }

    for (int neighbor : graph[node]) {
        if (!visited[neighbor]) {
            dfs(neighbor, graph, dist, forwardNode, visited, dest);
        }
        if (dist[neighbor] != -1 && dist[neighbor] + 1 > dist[node]) {
            dist[node] = dist[neighbor] + 1;
            forwardNode[node] = neighbor;
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<vector<int>> graph(n + 1);
    for (int i = 0; i < m; ++i) {
        int a, b;
        cin >> a >> b;
        graph[a].emplace_back(b);
    }

    vector<int> dist(n + 1, -1);
    vector<int> forwardNode(n + 1, -1);
    vector<bool> visited(n + 1, false);

    dfs(1, graph, dist, forwardNode, visited, n);

    if (dist[1] == -1) {
        cout << "IMPOSSIBLE" << endl;
        return 0;
    }

    cout << dist[1] << "\n";
    int index = 1;
    while (index != n) {
        cout << index << " ";
        index = forwardNode[index];
    }
    cout << n << endl;

    return 0;
}
